### PR TITLE
TFTP::Client retransmit lost data blocks on upload

### PR DESCRIPTION
**Pull request by amatus**
Retransmit data blocks until we receieve a matching ACK.

### CODE REVIEW

File: lib/rex/proto/tftp/client.rb [hunk 1]
```diff
@@ -302,21 +302,27 @@ def send_data(host,port)
...
-      res = self.server_sock.recvfrom(65535)
-      if res
-        code, type, msg = parse_tftp_response(res[0])
-        if code == 4
-          sent_blocks += 1
-          yield "Sent #{data_block.size} bytes in block #{sent_blocks}" if block_given?
-        else
-          if block_given?
-            yield "Got an unexpected response: Code:%d, Type:%d, Message:'%s'. Aborting." % [code, type, msg]
+      while true
```
Comment (wvu-r7) [inline lib/rex/proto/tftp/client.rb]: ```loop do``` would be good here.

**Comment by bcook-r7** [comment]
I ran into a lot of trouble just using the tftp client with or without this change. It doesn't work on OS X (UDP Rex socket issues). It indicates success when a real TFTP server sends an error packet. It indicates failure when the Rex TFTP server sends an OK. I'm going to need to understand how this works better to make it fully functional again.

**Comment by busterb** [comment]
I've been sitting on a few fixes for this in the mean time that might as well land. The main issue was the retries will spin 100% cpu if the remote server is not actually responding, so I added a max retry before giving up.

**Comment by bcook-r7** [comment]
6e146794a27d8e8d6545893c249f36497a7cbae2, b40c36688c2ee19ba60930d09e6631ba752425b3

**Comment by bcook-r7** [comment]
Thanks, #6001 sums up the remaining things to fix with TFTP.

**Pull request merged**